_posts/2016-05-10-finite-state-machines-with-akka.md

```diff
@@ -0,0 +1,206 @@
+
+After that we publish every item into Kafka. Remember that a single physical item can live in different containers, so that's why we talk about `items`. `ItemReportedProducer` returns a future and we can `pipe` it into the same actor. Thanks to that we can listen to failures of that future and make the process fail early.
+
+We subscribe the item into an [Akka Event Bus](http://doc.akka.io/docs/akka/2.4.4/java/event-bus.html), so this actor can react exclusively to its deleted items. Finally, we'll move the FSM into Active state including the state data of `ItemsToBeDeleted`.
+
+## Updating Data State in a FSM
+
+`ItemCensor` actor needs to wait until Items service finish deleting the items. Items service will publish some events into Kafka, and our Event Bus will be subscribed to that topic. `ItemCensor` is subscribed to only the items that he's interested to, and the Event Bus will send messages of type `ItemDeleted` to the actor.
+
+```scala
+  when(Active) {
+    case Event(ItemDeleted(item), currentItemsToBeDeleted@ItemsToBeDeleted(items)) =>
+      val newItemsToBeDeleted = items.filterNot(_ == item)
+      newItemsToBeDeleted.size match {
+        case 0 => finishWorkWith(CensorResult(Right()))
+        case _ => stay using currentItemsToBeDeleted.copy(items = newItemsToBeDeleted)
+      }
+  }
+```
+
+As soon as we get `ItemDeleted` messaged from the Event Bus we update the state data `ItemsToBeDeleted`. If we exhaust the items, then we can finish successfully with a [Right](http://www.scala-lang.org/api/rc2/scala/Either.html) message. Otherwise we `stay` in the same state with the new state data, waiting until new messages arrive.
+
+## Dealing with failure
+
+FSM in Akka allows you to capture messages that the actor received but no partial function matched it properly.
+
+```scala
+  whenUnhandled {
+    case Event(CensorTimeout, _) =>
+      finishWorkWith(CensorResult(Left(CensorTimeout("Censor timed out"))))
+
+    case Event(failure: Failure, _) =>
+      finishWorkWith(CensorResult(Left(CensorException(failure.cause.getMessage))))
+  }
+```
+`whenUnhandled` will try to match every handled message. If after 10 seconds the actor is still around a `CensorTimeout` message will be sent by `CensorTimer` so we can finish the work with the proper error case. If `itemReportedProducer` fails publishing an item to Kafka, this code will receive a `Failure` message as we piped that future into `self`.
+
+## Finishing the FSM
+
+The lifecycle of the FSM will be controlled from an outside actor, called coordinator. Whenever we want to finish the work of this FSM, we'll have to send a message to the coordinator:
+
+```scala
+  private def finishWorkWith(message: Any) = {
+    coordinator ! message
+    goto(Idle)
+  }
+```
+
+We don't need to go to `Idle` status, but as doing it is clearer to the reader that that actor is not on duty anymore.
```

Comment:
We don't need to go to `Idle` status, but doing it makes it clearer to the reader that the actor is not on duty anymore.